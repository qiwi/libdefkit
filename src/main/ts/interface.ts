/** @module @qiwi/libdefkit */
/** */

import {StdioOptions} from 'child_process'

export type ICmdInvokeOptions = {
  cmd: string
  args?: string[] | Record<string, any>,
  cwd?: string,
  silent?: boolean,
  stdio?: StdioOptions,
  closest?: boolean
}

export type ICliFlags = {
  cwd?: string
  entry?: string,
  dtsOut?: string,
  flowOut?: string,
  tsconfig?: string[]
}

export type IContext = ICliFlags & {
  cache: string
  cwd: string
  name: string,
  [key: string]: any
}

export type IExecPipe = (ctx: IContext) => IContext | void
