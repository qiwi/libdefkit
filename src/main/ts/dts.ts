/** @module @qiwi/libdefkit */

/** */
import path from 'node:path'
import fse from 'fs-extra'
import { globbySync } from 'globby'
import { generateDts } from 'tsc-dts-fix'
import { populateSync } from '@topoconfig/extends'

import { IContext, IExecPipe } from './interface.js'
import { findCommon } from './util.js'

export const generate = (
  tsconfig: string,
  name: string,
  cwd: string,
  entryPoints: Record<string, string> = {},
  ext: string,
): string => {
  const cfg = populateSync(fse.readJsonSync(tsconfig), {
    merge: {
      compilerOptions: 'merge',
    }
  })

  const {
    include = [],
    compilerOptions: {
      outDir = '' ,
      rootDir = findCommon(globbySync(include, {
        onlyFiles: true,
        absolute: true,
        cwd,
      }))
    } = {},
  } = cfg

  return generateDts({
    input: path.join(rootDir, 'index.ts'),
    strategy: 'bundle',
    pkgName: name,
    outDir,
    entryPoints,
    conceal: false,
    ext,
    cwd
  })['bundle.d.ts']
}

export const mixFiles = (files: string[], memo = ''): string =>
  files.reduce(
    (m: string, f: string) => m + "\n" + fse.readFileSync(f, { encoding: 'utf-8' }),
    memo,
  )

export const pipe: IExecPipe = (ctx): IContext => {
  const {
    name,
    entry,
    ext,
    tsconfig = [],
    customTypings = [],
    dtsOut,
    cwd,
  } = ctx

  let dts = tsconfig.map((cfg, i) =>
    generate(
      cfg,
      name,
      cwd,
      i === 0 ? {'.': entry ?? './index.ts'} : {},
      ext
    ),
  ).join('\n\n')

  dts = mixFiles(globbySync(customTypings, {
    onlyFiles: true,
    absolute: true,
    cwd,
  }), dts)

  fse.outputFileSync(dtsOut, dts)

  return { ...ctx, dts }
}
