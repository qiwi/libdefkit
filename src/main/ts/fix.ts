import {resolve} from 'path'
import { argv } from 'yargs'
import assert from 'assert'
import rep, {ReplaceInFileConfig, replaceInFileSync, ReplaceResult} from 'replace-in-file'
import {readFileSync} from 'fs'

const {dts, prefix} = argv
const DTS = resolve(dts as string)
const IMPORT_MAIN_PATTERN = /\timport main = require\('(.+)'\);/g
const IMPORT_MAIN_LINE_PATTERN = /^\timport main = require\('(.+)'\);$/
const BROKEN_MODULE_NAME = /(declare module '.+\/target\/es5\/)[^/]*\/src\/main\/index'.+/
const IMPORT = /import .+ from '(.+)'/g
const REFERENCE = /\/\/\/.+/

assert(!!dts, ' `dts` file path should be specified')

const dtsFile = readFileSync(DTS, 'utf-8')
const declaredModules = (dtsFile.match(/declare module '.*'/g) || []).map(v => v.slice(16, -1))

// console.log(declaredModules)
console.log(rep.sync)

const options: ReplaceInFileConfig = {
  files: DTS,
  from: [
    '\texport = main;',
    IMPORT_MAIN_PATTERN,
    BROKEN_MODULE_NAME,
    REFERENCE,
    /^\s*[\r\n]/gm,
    IMPORT
  ],
  to: [
    '',
    (line: string) => {
      const [, name] = IMPORT_MAIN_LINE_PATTERN.exec(line) || []
      return `	export * from '${name}';`
    },
    (line: string) => {
      const [, module] = BROKEN_MODULE_NAME.exec(line) || []
      return `${module}index' {`
    },
    '',
    '',
    (line: string) => {
      const re = /^(.+from ')([^']+)('.*)$/
      const [, pre, module, post] = re.exec(line) || []
      const name = declaredModules.includes(module)
        ? module
        : module.replace(prefix + '/', '')

      return `${pre}${name}${post}`
    }
  ],
}

const changes: ReplaceResult[] = replaceInFileSync(options);
console.log('Modified files:', changes);
