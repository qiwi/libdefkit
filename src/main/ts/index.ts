/** @module @qiwi/libdefkit */
/** */

export * from './rinf'
export * from './interface'
export * from './logger'
export * from './replacer'
