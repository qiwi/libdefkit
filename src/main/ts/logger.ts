import {ILogger} from '@qiwi/substrate'

export const logger: ILogger = console
