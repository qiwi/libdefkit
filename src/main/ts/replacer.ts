import {IReplacer, IReplacerFactory} from './interface'

export const replaceExportMain: IReplacer = {
  from: '\texport = main;',
  to: ''
}

export const replaceImportMain: IReplacer = {
  from: /\timport main = require\('(.+)'\);/g,
  to: (line: string) => {
    const [, name] = /^\timport main = require\('(.+)'\);$/.exec(line) || []
    return `	export * from '${name}';`
  }
}

export const replaceBrokenModulePrefix: IReplacer = (pattern => ({
  from: pattern,
  to: (line: string) => {
    const [, module] = pattern.exec(line) || []
    return `${module}index' {`
  }
}))(/(declare module '.+\/target\/es5\/)[^/]*\/src\/main\/index'.+/)

export const replaceModuleTypeRefs: IReplacer = {
  from: /\/\/\/.+/,
  to: ''
}

export const replaceEmptyLines: IReplacer = {
  from: /^\s*[\r\n]/gm,
  to: ''
}

export const replaceLocalModulesScope: IReplacerFactory = ({dtsData, prefix}) => {
  const declaredModules = (dtsData.match(/declare module '.*'/g) || []).map(v => v.slice(16, -1))

  return {
    from: /import .+ from '(.+)'/g,
    to: (line: string) => {
      const re = /^(.+from ')([^']+)('.*)$/
      const [, pre, module, post] = re.exec(line) || []
      const name = declaredModules.includes(module)
        ? module
        : module.replace(prefix + '/', '')

      return `${pre}${name}${post}`
    }
  }
}
