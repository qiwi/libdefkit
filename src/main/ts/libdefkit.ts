/** @module @qiwi/libdefkit */
/** */

import fs from 'fs-extra'
import { join } from 'node:path'
import { temporaryDirectory } from 'tempy'

import { pipe as dtsgen } from './dts'
import { ICliFlags, IContext, IExecPipe } from './interface'
import { invoke } from './util'

export const normalize: IExecPipe = (flags: ICliFlags): IContext => {
  const cwd = flags.cwd || process.cwd()
  const cache = temporaryDirectory()
  const name = fs.readJsonSync(join(cwd, 'package.json')).name
  const entry = flags.entry ?? `${name}/target/es6`
  const dtsOut = flags.dtsOut ?? join(cwd, 'typings', 'index.d.ts')
  const flowOut = flags.flowOut ?? join(cwd, 'flow-typed', 'index.flow.js')

  return { cache, ...flags, cwd, name, entry, dtsOut, flowOut }
}

export const clear: IExecPipe = (ctx) => {
  fs.emptyDirSync(ctx.cache)
}

export const flowgen: IExecPipe = ({ dtsOut, flowOut }): void => {
  if (dtsOut && flowOut) {
    invoke({
      cmd: 'flowgen',
      args: [dtsOut, '--output-file', flowOut],
      closest: true,
    })
  }
}

export const pipeline: IExecPipe[] = [normalize, clear, dtsgen, flowgen, clear]

export const execute = (flags: ICliFlags): IContext =>
  pipeline.reduce((ctx, pipe) => pipe(ctx) || ctx, flags as IContext)
