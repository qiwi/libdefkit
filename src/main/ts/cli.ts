#!/usr/bin/env node

import meow from 'meow'

import { execute } from './libdefkit'

const cli = meow(
  `
    Usage:
      libdefkit --out=some/path/
    Options
      --cwd, working dir path
      --tsconfig
      --customTypings, attach custom libdefs to d.ts bundle
      --entry, alias entry (<pkgName>/target/es5/index by default)
      --dtsOut, dts output (typings/index.d.ts by default)
      --flowOut, flow output (flow-typed/index.flow.js by default)
`,
  {
    importMeta: import.meta,
    flags: {
      cwd: {
        type: 'string',
      },
      dtsOut: {
        type: 'string',
      },
      flowOut: {
        type: 'string',
      },
      tsconfig: {
        type: 'string',
        isMultiple: true,
      },
      customTypings: {
        type: 'string',
        isMultiple: true,
      },
    },
  },
)

execute(cli.flags)
