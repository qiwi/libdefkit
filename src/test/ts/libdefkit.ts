import { join, resolve } from 'path'

import { execute } from '../../main/ts'

describe('libdefkit', () => {
  describe('#execute', () => {
    it('returns proper dts bundle', () => {
      const temp = resolve(__dirname, '..', 'temp')
      const dtsOut = join(temp, 'index.d.ts')
      const flowOut = join(temp, 'index.flow.js')
      const result = execute({
        tsconfig: ['tsconfig.es5.json', 'tsconfig.es6.json'],
        dtsOut,
        flowOut,
        customTypings: ['customTypings/**/*.d.ts']
      })

      const cwd = JSON.stringify(process.cwd()).slice(1, -1)
      const snap = JSON.parse(
        JSON.stringify(result).split(cwd).join('<cwd>').replace(/\\\\/g, '/'),
      )

      expect(snap).toMatchSnapshot()
    })
  })
})
