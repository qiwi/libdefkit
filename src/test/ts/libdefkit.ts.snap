// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`libdefkit #execute returns proper dts bundle 1`] = `
Object {
  "cache": "<cwd>/node_modules/.cache/@qiwi/libdefkit",
  "cwd": "<cwd>",
  "dts": "// Generated by dts-bundle v0.7.4-beta.1
// Dependencies for this module:
//   ../../../../../../../child_process

declare module '@qiwi/libdefkit/target/es6' {
    /** */
    export * from '@qiwi/libdefkit/target/es6/libdefkit';
}

declare module '@qiwi/libdefkit/target/es6/libdefkit' {
    /** */
    import { ICliFlags, IContext, IExecPipe } from '@qiwi/libdefkit/target/es6/interface';
    export const normalize: IExecPipe;
    export const clear: IExecPipe;
    export const flowgen: IExecPipe;
    export const pipeline: IExecPipe[];
    export const execute: (flags: ICliFlags) => IContext;
}

declare module '@qiwi/libdefkit/target/es6/interface' {
    /** */
    import { StdioOptions } from 'child_process';
    export type ICmdInvokeOptions = {
        cmd: string;
        args?: string[] | Record<string, any>;
        cwd?: string;
        silent?: boolean;
        stdio?: StdioOptions;
        closest?: boolean;
    };
    export type ICliFlags = {
        cwd?: string;
        entry?: string;
        dtsOut?: string;
        flowOut?: string;
        tsconfig?: string[];
    };
    export type IContext = ICliFlags & {
        cache: string;
        cwd: string;
        name: string;
        [key: string]: any;
    };
    export type IExecPipe = (ctx: IContext) => IContext | void;
}

// Generated by dts-bundle v0.7.4-beta.1
// Dependencies for this module:
//   ../../../../../../../child_process

declare module '@qiwi/libdefkit/target/es5' {
    /** */
    export * from '@qiwi/libdefkit/target/es5/libdefkit';
}

declare module '@qiwi/libdefkit/target/es5/libdefkit' {
    /** */
    import { ICliFlags, IContext, IExecPipe } from '@qiwi/libdefkit/target/es5/interface';
    export const normalize: IExecPipe;
    export const clear: IExecPipe;
    export const flowgen: IExecPipe;
    export const pipeline: IExecPipe[];
    export const execute: (flags: ICliFlags) => IContext;
}

declare module '@qiwi/libdefkit/target/es5/interface' {
    /** */
    import { StdioOptions } from 'child_process';
    export type ICmdInvokeOptions = {
        cmd: string;
        args?: string[] | Record<string, any>;
        cwd?: string;
        silent?: boolean;
        stdio?: StdioOptions;
        closest?: boolean;
    };
    export type ICliFlags = {
        cwd?: string;
        entry?: string;
        dtsOut?: string;
        flowOut?: string;
        tsconfig?: string[];
    };
    export type IContext = ICliFlags & {
        cache: string;
        cwd: string;
        name: string;
        [key: string]: any;
    };
    export type IExecPipe = (ctx: IContext) => IContext | void;
}

declare module '@qiwi/libdefkit' {
  /** */
  export * from '@qiwi/libdefkit/target/es5';
}
",
  "dtsOut": "<cwd>/src/test/temp/index.d.ts",
  "entry": "@qiwi/libdefkit/target/es5",
  "flowOut": "<cwd>/src/test/temp/index.flow.js",
  "name": "@qiwi/libdefkit",
  "tsconfig": Array [
    "tsconfig.es5.json",
    "tsconfig.es6.json",
  ],
}
`;
